reach/reach.py

```diff
@@ -68,49 +83,70 @@ def __init__(self, vectors, items, name="", unk_index=None):
                 "order."
             )
 
-        self.items = {w: idx for idx, w in enumerate(items)}
-        self.indices = {v: k for k, v in self.items.items()}
-
+        self._items: Dict[str, int] = {w: idx for idx, w in enumerate(items)}
+        self._indices: Dict[int, str] = {idx: w for w, idx in self.items.items()}
         self.vectors = np.asarray(vectors)
         self.unk_index = unk_index
-
         self.name = name
 
     @property
-    def size(self):
+    def items(self) -> Dict[str, int]:
+        return self._items
+
+    @property
+    def indices(self) -> Dict[int, str]:
+        return self._indices
+
+    def __len__(self) -> int:
+        return len(self.items)
+
+    @property
+    def sorted_items(self) -> List[str]:
+        items: List[str] = [
+            item for item, _ in sorted(self.items.items(), key=lambda x: x[1])
+        ]
+        return items
+
+    @property
+    def size(self) -> int:
         return self.vectors.shape[1]
 
     @property
-    def vectors(self):
+    def vectors(self) -> np.ndarray:
         return self._vectors
 
     @vectors.setter
-    def vectors(self, x):
-        x = np.array(x)
-        assert np.ndim(x) == 2 and x.shape[0] == len(self.items)
+    def vectors(self, x: Matrix) -> None:
+        x = np.asarray(x)
+        if not np.ndim(x) == 2:
+            raise ValueError(f"Your array does not have 2 dimensions: {np.ndim(x)}")
+        if not x.shape[0] == len(self.items):
+            raise ValueError(
+                f"Your array does not have the correct length, got {x.shape[0]}, expected {len(self.items)}"
+            )
         self._vectors = x
         # Make sure norm vectors is updated.
         if hasattr(self, "_norm_vectors"):
             self._norm_vectors = self.normalize(x)
```

Comment:
Ik snap deze twee niet zo goed, je definieert 2x vectors een keer als property en 1 keer als vectors.setter, wat doet dat? In de init wordt ook al self.vectors gezet volgens mij

Reply:
The init sets self.__vectors. The .setter is the setter for the property vectors and defines what happens when you set .vectors. .vectors defines the property as usual